# Module 2 Lab - Probability

## Directions

**Failure to follow the directions will result in a "0"**

The due dates for each are indicated in the Syllabus and the course calendar. If anything is unclear, please email EN605.448@gmail.com the official email for the course or ask questions in the Lab discussion area on Blackboard.

The Labs also present technical material that augments the lectures and "book".  You should read through the entire lab at the start of each module.

### General Instructions

1.  You will be submitting your assignment to Blackboard. If there are no accompanying files, you should submit *only* your notebook and it should be named using *only* your JHED id: fsmith79.ipynb for example if your JHED id were "fsmith79". If the assignment requires additional files, you should name the *folder/directory* your JHED id and put all items in that folder/directory, ZIP it up (only ZIP...no other compression), and submit it to Blackboard.
    
    * do **not** use absolute paths in your notebooks. All resources should appear in the same directory as the rest of your assignments.
    * the directory **must** be named your JHED id and **only** your JHED id.
    
2. Data Science is as much about what you write (communicating) as the code you execute (researching). In many places, you will be required to execute code and discuss both the purpose and the result. Additionally, Data Science is about reproducibility and transparency. This includes good communication with your team and possibly with yourself. Therefore, you must show **all** work.

3. Avail yourself of the Markdown/Codecell nature of the notebook. If you don't know about Markdown, look it up. Your notebooks should not look like ransom notes. Don't make everything bold. Clearly indicate what question you are answering.

4. Submit a cleanly executed notebook. It should say `In [1]` for the first codecell and increase by 1 throughout.

## Manipulating and Interpreting Probability

Given the following *joint probability distribution*, $P(A,B)$, for $A$ and $B$,


|    | a1   | a2   |
|----|------|------|

| **b1** | 0.37 | 0.16 |

| **b2** | 0.23 | ?    |

Answer the following questions.

**1\. What is $P(A=a2, B=b2)$?**

In [1]:
print("P(A=a2,B=b2) = 1 - 0.37  - 0.16 - 0.23 = %.2f" %(1 - 0.37 - 0.16 - 0.23))

P(A=a2,B=b2) = 1 - 0.37  - 0.16 - 0.23 = 0.24


In [ ]:
P = {"a1":{"b1":0.37,"b2"=0.23},"a2":{"b1":0.,"b2"=0.05}}

**2\. If I observe events from this probability distribution, what is the probability of seeing (a1, b1) then (a2, b2)?**

In [2]:
print("P((a1,b1)then(a2,b2)) = P(a1,b1) * P(a2,b2) = 0.37 * 0.24= %.2f" %(0.37*0.24))

P((a1,b1)then(a2,b2)) = P(a1,b1) * P(a2,b2) = 0.37 * 0.24= 0.09


**3\. Calculate the marginal probability distribution, $P(A)$.**

In [3]:
print("P(a1) = P(a1,b1) + P(a1,b2) = 0.37 + 0.23 = %.2f" %(0.37+0.23))
print("P(a2) = P(a2,b1) + P(a2,b2) = 0.16 + 0.24 = %.2f" %(0.16+0.24))

P(a1) = P(a1,b1) + P(a1,b2) = 0.37 + 0.23 = 0.60
P(a2) = P(a2,b1) + P(a2,b2) = 0.16 + 0.24 = 0.40


**4\. Calculate the marginal probability distribution, $P(B)$.**

In [4]:
print("P(b1) = P(b1,a1) + P(b1,a2) = 0.37 + 0.23 = %.2f" %(0.37+0.16))
print("P(b2) = P(b2,a1) + P(b2,a2) = 0.23 + 0.24 = %.2f" %(0.23+0.24))

P(b1) = P(b1,a1) + P(b1,a2) = 0.37 + 0.23 = 0.53
P(b2) = P(b2,a1) + P(b2,a2) = 0.23 + 0.24 = 0.47


**5\. Calculate the conditional probability distribution, $P(A|B)$.**

In [5]:
print("P(a1|b1) = P(a1,b1)/P(b1) = 0.37/0.53 = %.2f" %(0.37/0.53))
print("P(a1|b2) = P(a1,b2)/P(b2) = 0.23/0.47 = %.2f" %(0.23/0.47))
print("P(a2|b1) = P(a2,b1)/P(b1) = 0.16/0.53 = %.2f" %(0.16/0.53))
print("P(a2|b2) = P(a2,b2)/P(b2) = 0.0.24/0.47 = %.2f" %(0.24/0.47))

P(a1|b1) = P(a1,b1)/P(b1) = 0.37/0.53 = 0.70
P(a1|b2) = P(a1,b2)/P(b2) = 0.23/0.47 = 0.49
P(a2|b1) = P(a2,b1)/P(b1) = 0.16/0.53 = 0.30
P(a2|b2) = P(a2,b2)/P(b2) = 0.0.24/0.47 = 0.51


**6\. Calculate the conditional probability distribution, $P(B|A)$.**

In [6]:
print("P(b1|a1) = P(b1,a1)/P(a1) = 0.37/0.60 = %.2f" %(0.37/0.60))
print("P(b1|a2) = P(b1,a2)/P(a2) = 0.16/0.40 = %.2f" %(0.16/0.40))
print("P(b2|a1) = P(b2,a1)/P(a1) = 0.23/0.60 = %.2f" %(0.23/0.60))
print("P(b2|a2) = P(b2,a2)/P(a2) = 0.24/0.40 = %.2f" %(0.24/0.40))


P(b1|a1) = P(b1,a1)/P(a1) = 0.37/0.60 = 0.62
P(b1|a2) = P(b1,a2)/P(a2) = 0.16/0.40 = 0.40
P(b2|a1) = P(b2,a1)/P(a1) = 0.23/0.60 = 0.38
P(b2|a2) = P(b2,a2)/P(a2) = 0.24/0.40 = 0.60


**7\. Does $P(A|B) = P(B|A)$? What do we call the belief that these are always equal?**

Based on answers to Question 5 and 6, P(A|B) does not equal to P(B|A)
The belief that these are always equal is called inverse probability fallacy


**8\. Does $P(A) = P(A|B)$? What does that mean about the independence of $A$ and $B$?**

Based on answers to Question 3 and 5

P(a1)=0.60 whereas P(a1|b1)=0.70 and p(a1|b2)=0.49

P(a2)=0.40 whereas P(a2|b1)=0.53 and p(a2|b2)=0.47

Therefore, P(A) does not euqal to P(A|B) and they are not independent

**9\. Using $P(A)$, $P(B|A)$, $P(B)$ from above, calculate,**

$P(A|B) = \frac{P(B|A)P(A)}{P(B)}$

Does it match your previous calculation for $P(A|B)$?

In [7]:
print("Yes, all calculations match previous results. See below for details:")
print("P(a1|b1) = (P(b1|a1)*P(a1))/P(b1) = (0.62*0.60)/0.53 = %.2f" %((0.62*0.60)/0.53))
print("P(a1|b2) = (P(b2|a1)*P(a1))/P(b2) = (0.38*0.60)/0.47 = %.2f" %((0.38*0.60)/0.47))
print("P(a2|b1) = (P(b1|a2)*P(a2))/P(b1) = (0.40*0.40)/0.53 = %.2f" %((0.40*0.40)/0.53))
print("P(a2|b2) = (P(b2|a2)*P(a2))/P(b2) = (0.60*0.40)/0.47 = %.2f" %((0.60*0.40)/0.47))


Yes, all calculations match previous results. See below for details:
P(a1|b1) = (P(b1|a1)*P(a1))/P(b1) = (0.62*0.60)/0.53 = 0.70
P(a1|b2) = (P(b2|a1)*P(a1))/P(b2) = (0.38*0.60)/0.47 = 0.49
P(a2|b1) = (P(b1|a2)*P(a2))/P(b1) = (0.40*0.40)/0.53 = 0.30
P(a2|b2) = (P(b2|a2)*P(a2))/P(b2) = (0.60*0.40)/0.47 = 0.51


**10\. If we let A = H (some condition, characteristic, hypothesis) and B = D (some data, evidence, a test result), then how do we interpret each of the following: $P(H)$, $P(D)$, $P(H|D)$, $P(D|H)$, $P(H, D)$?**

Referring to **Fundamental Chapter 5, Secsion 5.5.5**, each of these should be interpreted as:
- P(H) = prior probability
- P(D) = normalizer
- P(H|D) = posterior probability
- P(D|H) = likelihood
- P(H,D) = joint probability

## Bayes Rule

Bayes Rule will be an important part of our toolset in the weeks to come, especially in terms of Bayesian Inference. Work through the following problems in Bayes Rule.

### Problem 1

You might be interested in finding out a patient’s probability of having liver disease if they are an alcoholic. “Being an alcoholic” is the test (kind of like a litmus test) for liver disease.

Let `A` be the presence or absence of liver disease (`a` is that they have it; `~a`, "not a", is that they don't). Past data tells you that 10% of patients entering your clinic have liver disease.

Let `B` be alcoholic (`b`) or not alcoholic (`~b`). 5% of the clinic’s patients are alcoholics.

You might also know that among those patients diagnosed with liver disease, 7% are alcoholics.

What is P(A|B)? (Remember this is a number of probability distributions, not just a probability).

Given P(b)=0.05, P(~b)=0.95, P(a)=0.10, p(~a)=0.90, p(b|a)=0.07,p(~b|a)=0.93

Using Bayes rule, we can get P(A|B) as:


In [8]:
print("P(a|b) = (P(b|a)*P(a))/P(b) = 0.07*0.10/0.05 = %.2f" %((0.07*0.10)/0.05))
print("P(~a|b) = 1-P(a|b) = %.2f" %(1-(0.07*0.10)/0.05))
print("P(a|~b) = (P(~b|a)*P(a))/P(~b) = 0.93*0.10/0.95 = %.2f" %((0.93*0.10)/0.95))
print("P(~a|~b) = 1-P(~b|a) = %.2f" %(1-((0.93*0.10)/0.95)))
print("P(b|a) = P(a|b)*P(b)/P(a) = 0.14*0.05/0.10 = %.2f" %((0.14*0.05)/0.10))
print("P(~b|a) = 1-P(b|a) = 1-0.14*0.05/0.10 = %.2f" %(1-0.14*0.05/0.10))     
print("P(b|~a) = P(~a|b)*P(b)/P(~a) = 0.86*0.05/0.90 = %.2f" %(0.86*0.05/0.90))
print("P(~b|~a) = 1-P(b|~a) = %.2f" %(1-0.86*0.05/0.90))


P(a|b) = (P(b|a)*P(a))/P(b) = 0.07*0.10/0.05 = 0.14
P(~a|b) = 1-P(a|b) = 0.86
P(a|~b) = (P(~b|a)*P(a))/P(~b) = 0.93*0.10/0.95 = 0.10
P(~a|~b) = 1-P(~b|a) = 0.90
P(b|a) = P(a|b)*P(b)/P(a) = 0.14*0.05/0.10 = 0.07
P(~b|a) = 1-P(b|a) = 1-0.14*0.05/0.10 = 0.93
P(b|~a) = P(~a|b)*P(b)/P(~a) = 0.86*0.05/0.90 = 0.05
P(~b|~a) = 1-P(b|~a) = 0.95


### Problem 2

In a particular pain clinic, 10% of patients are prescribed narcotic pain killers. Overall, five percent of the clinic’s patients are addicted to narcotics (including pain killers and illegal substances). Out of all the people prescribed pain pills, 8% are addicts. If a patient is an addict, what is the probability that they will be prescribed pain pills?

Let N be the presence or absence of pain killer prescribtion (n is that they have it; ~n, "not n", is that they don't)

Let A be addict (a) or not addict (~a)

Given P(n)=0.10, P(a)=0.05, P(a|n)=0.08

Using Bayes rule, we can get:


In [9]:
print("P(n|a) = (P(a|n)*P(n))/P(a) = (0.08*0.10)/0.05 = %.2f" %((0.08*0.10)/0.05))

P(n|a) = (P(a|n)*P(n))/P(a) = (0.08*0.10)/0.05 = 0.16


Therefore, if a patient is an addict, the probability that they will be prescribed pain pills is 16%